// ngx-gallery-9/esm5/lib/ngx-gallery-bullets/ngx-gallery-bullets.component.js
import { __decorate } from "tslib";
import { Component, Input, EventEmitter, Output } from '@angular/core';
var NgxGalleryBulletsComponent = /** @class */ (function () {
    function NgxGalleryBulletsComponent() {
        this.active = 0;
        this.onChange = new EventEmitter();
    }
    NgxGalleryBulletsComponent.prototype.getBullets = function () {
        return Array(this.count);
    };
    NgxGalleryBulletsComponent.prototype.handleChange = function (event, index) {
        this.onChange.emit(index);
    };
    __decorate([
        Input()
    ], NgxGalleryBulletsComponent.prototype, "count", void 0);
    __decorate([
        Input()
    ], NgxGalleryBulletsComponent.prototype, "active", void 0);
    __decorate([
        Output()
    ], NgxGalleryBulletsComponent.prototype, "onChange", void 0);
    NgxGalleryBulletsComponent = __decorate([
        Component({
            selector: 'ngx-gallery-bullets',
            template: "<div class=\"ngx-gallery-bullet\" *ngFor=\"let bullet of getBullets(); let i = index;\" (click)=\"handleChange($event, i)\" [ngClass]=\"{ 'ngx-gallery-active': i == active }\"></div>",
            styles: [":host{position:absolute;z-index:2000;display:-webkit-inline-box;display:inline-flex;left:50%;-webkit-transform:translateX(-50%);transform:translateX(-50%);bottom:0;padding:10px}.ngx-gallery-bullet{width:10px;height:10px;border-radius:50%;cursor:pointer;background:#fff}.ngx-gallery-bullet:not(:first-child){margin-left:5px}.ngx-gallery-bullet.ngx-gallery-active,.ngx-gallery-bullet:hover{background:#000}"]
        })
    ], NgxGalleryBulletsComponent);
    return NgxGalleryBulletsComponent;
}());
export { NgxGalleryBulletsComponent };
